cjdns-snode/src/server/webserver.rs

```diff
@@ -72,45 +101,512 @@ mod handlers {
+    use super::utils::{to_label64_string, to_ser_enc_scheme};
 
     pub(super) async fn handle_info(server: Arc<Server>) -> Result<impl Reply, Infallible> {
         let reply = InfoReply {
             peers_info: server.peers.get_info(),
             nodes_count: server.nodes.count()
         };
-        Ok(reply::json(&reply))
+        Ok(reply_json(&reply))
     }
 
     pub(super) async fn handle_debug_node(ip6: String, server: Arc<Server>) -> Result<StatusCode, Rejection> {
-        let ip = CJDNS_IP6::try_from(ip6.as_str()).map_err(|_| warp::reject::custom(WebServerError::BadIP6Address))?;
+        let ip = CJDNS_IP6::try_from(ip6.as_str()).map_err(|_| warp::reject::custom(WebServerError::BadIP6Address(ip6)))?;
         server.mut_state.lock().debug_node = Some(ip);
         return Ok(StatusCode::OK);
     }
 
     pub(super) async fn handle_dump(server: Arc<Server>) -> Result<Vec<u8>, Infallible> {
         Ok(server.nodes.anns_dump())
     }
+
+    pub(super) async fn handle_path(src: String, tar: String, server: Arc<Server>) -> Result<impl Reply, Rejection> {
+        let src_ip = CJDNS_IP6::try_from(src.as_str()).map_err(|_| warp::reject::custom(WebServerError::BadIP6Address(src)))?;
+        let tar_ip = CJDNS_IP6::try_from(tar.as_str()).map_err(|_| warp::reject::custom(WebServerError::BadIP6Address(tar)))?;
+        let src = server.nodes.by_ip(&src_ip);
+        let tar = server.nodes.by_ip(&tar_ip);
+        warn!("http getRoute req {} {}", src_ip, tar_ip);
+        if src.is_none() { return Ok("src not found".to_string()); }
+        if tar.is_none() { return Ok("tar not found".to_string()); }
+        match get_route(server.clone(), src, tar) {
+            Ok(r) => Ok(r.label.to_string()),
+            Err(e) => Ok(format!("{:?}", e))
+        } 
+    }
+
+    pub(super) async fn handle_ni_with_ip(ip6: String, server: Arc<Server>) -> Result<impl Reply, Infallible> {
+        let ip6 = CJDNS_IP6::try_from(ip6.as_str());
+        if let Ok(ip6) = ip6 {
+            if let Some(node) = server.nodes.by_ip(&ip6) {
+                let node = SerNode(node);
+                return Ok(reply_json(&NiReply::NiIp{node}));
+            }
+        }
+        return Ok(reply_json(&NiReply::Empty{}))
+    }
+
+    pub(super) async fn handle_ni_empty(server: Arc<Server>) -> Result<impl Reply, Infallible> {
+        let mut total_ann = 0;
+        let mut resets = 0;
+        let nodes: Vec<_> = server.nodes.all_ips().iter().map(|ip6| {
+            let node = server.nodes.by_ip(ip6).expect("internal error: no node with provided ip");
+            let node_mut = node.mut_state.read();
+            let announcements = node_mut.announcements.len();
+            total_ann += announcements;
+            let rst = node_mut.reset_msg.is_some() && node_mut.announcements.iter().any(|ann| Some(ann) == node_mut.reset_msg.as_ref());
```

Comment:
Is it correct??

Reply:
should check whether it's not in collection, right. For example: 
`!node_mut.announcements.iter().any(|ann| Some(ann) == node_mut.reset_msg.as_ref())`